winch(x64): Initial implementation for function calls

saulecabrera opened the pull request:
This change adds the main building blocks for calling locally defined
functions. Support for function imports will be added iteratively after this
change lands and once trampolines are supported. Which is also when 
we plan to introduce support for the `Wasmtime*` calling conventions.

To support function calls, this change introduces the following functionality to
the MacroAssembler:

* `pop` to pop the machine stack into a given register, which in the case of
this change, translates to the x64 pop instruction.

* `call` to a emit a call to locally defined functions.

* `address_from_sp` to construct memory addresses with the SP as a base.

* `free_stack` to emit the necessary instrunctions to claim stack space.

The heavy lifting of setting up and emitting the function call is done through
the implementation of `FnCall`.

<!--

Please ensure that the following steps are all taken care of before submitting
the PR.

- [ ] This has been discussed in issue #..., or if not, please tell us why
  here.
- [ ] A short description of what this does, why it is needed; if the
  description becomes long, the matter should probably be discussed in an issue
  first.
- [ ] This PR contains test cases, if meaningful.
- [ ] A reviewer from the core maintainer team has been assigned for this PR.
  If you don't know who could review this, please indicate so. The list of
  suggested reviewers on the right can help you.

Please ensure all communication adheres to the [code of
conduct](https://github.com/bytecodealliance/wasmtime/blob/master/CODE_OF_CONDUCT.md).
-->

saulecabrera commented:
Oh It seems I messed up the branches and that's why https://github.com/bytecodealliance/wasmtime/pull/6067/commits/03637c95fa75767a9c4cfedd638cb1f45016053d which is unrelated to this change appears. I'll fix that.

saulecabrera commented on winch/codegen/src/codegen/context.rs:
```diff
@@ -60,58 +61,66 @@ impl<'a> CodeGenContext<'a> {
+    /// it isn't already one; spilling if there are no registers
+    /// available.  Optionally the caller may specify a specific
+    /// destination register.
+    pub fn pop_to_reg<M: MacroAssembler>(
```
The work by `pop_to_reg` and `pop_to_named_reg` was very similar, so I realized I could remove some repetition here through this refactor.

cfallin commented on winch/codegen/src/codegen/call.rs:
```diff
@@ -0,0 +1,144 @@
+
+        let spilled_regs = match callee_params.len() {
+            0 => {
+                let _ = context.spill_regs_in(masm, ..);
```
Does this need to be a special case? It seems that if instead the `spill_regs_in` method accepted an empty range, this case would work with the code below...

saulecabrera replied:
I was not able to find a better way to do this. In this case we still want to spill everything, to save any live registers, but we are no interested in the count returned by `spill_regs_and_count_memory_in`, since they are not part of the function arguments so we can't claim that memory after the function call, so that's why I'm handling `0` in this way.

cfallin commented on winch/codegen/src/codegen/call.rs:
```diff
@@ -0,0 +1,144 @@
+use super::CodeGenContext;
```
A top-level module comment describing the calling convention would be really helpful here! In particular, which registers are clobbered or saved, and where arguments are passed.

saulecabrera replied:
Agreed! Added the documentation in the abi module, and added a reference to it from here.

cfallin commented on winch/filetests/filetests/x64/call/params.wat:
```diff
@@ -0,0 +1,99 @@
+;;! target = "x86_64"
```
It might be good to have a test initially that exercises the spilling behavior (here we can't because `$add` is called in tail position). Perhaps call it multiple times, with some values on the stack that live "under" the call?

saulecabrera replied:
Fixed! Thanks.

cfallin reviewed (approved):
Looks generally great, thanks! Just some clarification-comment thoughts and possibly another test case.

saulecabrera commented:
@cfallin thanks for the review and sorry it took me a bit to get back to this. 

Good call on introducing a more complex test -- I was able to uncover a bug regarding addresses in the spill behaviour. I've introduced a fix and an explanation in https://github.com/bytecodealliance/wasmtime/pull/6067/commits/5e058d21165477cd4ae1d1257c7f7f218d604f50. Given that I introduced some new changes, I'd appreciate if you could take another look. Thanks!

saulecabrera commented on winch/codegen/src/codegen/call.rs:
```diff
@@ -0,0 +1,144 @@
+                assert!(stack.len() >= callee_params.len());
+                let partition = stack.len() - callee_params.len();
+                let _ = context.spill_regs_in(masm, 0..partition);
+                context.spill_regs_in(masm, partition..)
```
Fixed!

cfallin commented on winch/codegen/src/codegen/call.rs:
```diff
@@ -11,15 +13,24 @@ pub(crate) struct FnCall<'a> {
+    /// and thus we track this space to claim it once the function
+    /// call has ended. This field is analogous to
+    /// `spill_stack_space`, but it's kept separate for clarity and
+    /// debugging purposes.
```
I definitely appreciate the docs here, but a few things are slightly unclear to me at least:

- The terminology of "memory entries" is confusing -- does this just mean "values on the stack"?
- The "That is, any memory entries ... " sentence is really hard to parse. I *think* what this is saying is: values already on the stack that are args?
- "Claim" means "reclaim" here?

Is it the case that `memory_consumed + spill_stack_space == total_arg_stack_space`? If so, could we use that alone to pop args off the stack? Fundamentally it seems like we shouldn't have to know where the args originally came from when we're cleaning up.

saulecabrera replied:
> Is it the case that memory_consumed + spill_stack_space == total_arg_stack_space?

I see where you're coming from and I agree that it's a bit confusing, I think I can make it so that it's more clear and so that we have to deal with less terms.

But to give a bit more context:

* total_arg_stack_space is the total space that needs to be explicitly allocated at the callsite, and covers: the stack space needed for arguments plus any alignment needed for the caller's frame. 

* memory_consumed, represents the memory used by any `Memory` entries existing in the value stack at the moment of assigning those entries to the function arguments (in the case of a register argument, that memory will be loaded to the register used as argument, and in the case of a stack argument, that memory will be loaded to the scratch register and then stored to the particular argument slot). We could emit a `pop` instruction when doing this assignment and will potentially simplify things here in terms of what we are tracking, but I think it's more efficient to track the amount of memory used and reclaim it in the end.

* spill_stack_space is the space created to save any live registers.

Does that clarify things to you? 

As I write this, I believe we can simplify this by tracking all the memory used in a single field. Initially I thought that separating where things come from would make the code more understandable and easier to reason about, but maybe the opposite is better.

saulecabrera replied:
I've simplified how we track these values in https://github.com/bytecodealliance/wasmtime/pull/6067/commits/2e5372c8ffb430e75442ef33cc5cd7a3b408800e and also added more extensive documentation. Hopefully this clarifies the code.

cfallin commented on winch/codegen/src/codegen/context.rs:
```diff
@@ -182,31 +182,35 @@ impl<'a> CodeGenContext<'a> {
     ///
-    /// Returns the amount of registers spilled for the given range.
-    pub fn spill_regs_in<M, R>(&mut self, masm: &mut M, range: R) -> u32
+    /// Returns the amount of registers spilled and the amount of
```
"Amount of registers" -> "number of registers"? (register is a count-noun so the "amount of" reads ambiguously to me: in bytes, or whole registers, or ...)

cfallin commented on winch/codegen/src/codegen/call.rs:
```diff
@@ -87,9 +125,9 @@ impl<'a> FnCall<'a> {
         self.assign_args(context, masm, <A as ABI>::scratch_reg());
         masm.call(callee);
-        masm.free_stack(self.spill_stack_space + self.total_arg_stack_space);
+        masm.free_stack(self.spill_stack_space + self.total_arg_stack_space + self.memory_consumed);
```
I'm a little confused here: shouldn't we be popping just `total_arg_stack_space`, without also popping `memory_consumed` and `spill_stack_space`? It seems like this is double-counting, unless I'm misunderstanding what the above amounts mean.

(Labeling portions of the stack in the figure might help clarify, too, or else noting the state of hte stack at points in the sequence above)

saulecabrera replied:
Agreed, I've added a figure of how the stack looks like at each point in time in https://github.com/bytecodealliance/wasmtime/commit/2e5372c8ffb430e75442ef33cc5cd7a3b408800e, hopefully this helps clarify!

cfallin reviewed (approved):
This is clearer now with the updated comments and after our discussion today -- thanks very much for the patience! LGTM.
